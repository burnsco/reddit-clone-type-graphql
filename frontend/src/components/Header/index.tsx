import {
  Button,
  Flex,
  Menu,
  MenuButton,
  MenuGroup,
  MenuItem,
  MenuList
} from '@chakra-ui/core'
import { TestModal } from '@components/Modal'
import { useLogoutMutation, useMeQuery } from '@generated/graphql'
import { useRouter } from 'next/router'
import * as React from 'react'
import { ColorModeToggle } from '../layout/ColorModeToggle'

const Header: React.FC = () => {
  const router = useRouter()
  const { data, loading } = useMeQuery({ ssr: false })
  const [logout, { client }] = useLogoutMutation()
  if (loading) return null

  return (
    <Flex
      as='header'
      minW='100%'
      mb={8}
      align='center'
      borderBottom='1px'
      borderBottomStyle='solid'
      borderBottomColor='gray.200'
      height='60px'
    >
      <Flex minW='100%' as='nav' justify='space-evenly'>
        {data && data?.me?.username ? (
          <Menu>
            <TestModal />

            {/* SUBMIT MENU */}
            <Menu>
              <MenuButton as={Button}>Create</MenuButton>
              <MenuList>
                <MenuItem onClick={() => router.push('/create-post')}>
                  Post
                </MenuItem>
                <MenuItem onClick={() => router.push('/create-subreddit')}>
                  Subreddit
                </MenuItem>
              </MenuList>
            </Menu>

            {/* PROFILE MENU  */}
            <Menu>
              <MenuButton as={Button}>{data.me.username}</MenuButton>
              <MenuList>
                <MenuGroup title='Settings'>
                  <MenuItem onClick={() => router.push('/settings/profile')}>
                    Profile
                  </MenuItem>
                  <MenuItem onClick={() => router.push('/settings/account')}>
                    Account
                  </MenuItem>
                  <MenuItem onClick={() => router.push('/settings')}>
                    General
                  </MenuItem>
                </MenuGroup>
                <MenuItem
                  mr={2}
                  color='red.500'
                  onClick={async () => {
                    await logout()
                    await client.resetStore()
                  }}
                >
                  {' '}
                  Logout
                </MenuItem>
              </MenuList>
            </Menu>
          </Menu>
        ) : (
          <Menu>
            <MenuButton
              variant='outline'
              mr={2}
              as={Button}
              onClick={() => router.push('/register')}
            >
              Register
            </MenuButton>
            <MenuButton
              mr={2}
              as={Button}
              onClick={() => router.push('/login')}
            >
              Login
            </MenuButton>
          </Menu>
        )}
        <ColorModeToggle mr={2} />
      </Flex>
    </Flex>
  )
}

export default Header
