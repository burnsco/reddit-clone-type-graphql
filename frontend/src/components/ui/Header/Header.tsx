import {
  chakra,
  Flex,
  IconButton,
  useColorMode,
  useColorModeValue,
  useSafeLayoutEffect,
  useToast
} from "@chakra-ui/react"
import { FaMoon, FaSun } from "react-icons/fa"
import useNewUserNotification from "../../../hooks/useNewUserNotify"
import NavSection from "./Center"
import LogoSection from "./Left"
import MenuIconsSection from "./Right"

const HeaderContent = () => {
  const { toggleColorMode: toggleMode } = useColorMode()
  const text = useColorModeValue("dark", "light")
  const SwitchIcon = useColorModeValue(FaMoon, FaSun)
  return (
    <Flex w="100%" h="100%" px="6" align="center" justify="space-between">
      <LogoSection />
      <NavSection />

      <MenuIconsSection />
      <IconButton
        size="md"
        fontSize="lg"
        aria-label={`Switch to ${text} mode`}
        variant="ghost"
        color="current"
        ml="3"
        onClick={toggleMode}
        icon={<SwitchIcon />}
      />
    </Flex>
  )
}

const Header = () => {
  const toast = useToast()
  const user = useNewUserNotification()

  useSafeLayoutEffect(() => {
    if (user) {
      toast({
        position: "bottom-left",
        title: `User "${user}"`,
        description: "Has just joined the community!",
        status: "warning",
        duration: 5000,
        isClosable: true
      })
    }
  }, [user])

  return (
    <chakra.header
      pos="fixed"
      top="0"
      zIndex="1"
      bg={useColorModeValue("white", "#202020")}
      left="0"
      right="0"
      boxShadow={useColorModeValue("md", "dark-lg")}
      width="full"
    >
      <chakra.div height="3.5rem" mx="auto" maxW="1200px">
        <HeaderContent />
      </chakra.div>
    </chakra.header>
  )
}

export default Header
