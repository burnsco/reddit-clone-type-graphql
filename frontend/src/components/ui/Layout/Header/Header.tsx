import {
  chakra,
  Flex,
  IconButton,
  useColorMode,
  useColorModeValue
} from "@chakra-ui/core"
import { FaMoon, FaSun } from "react-icons/fa"
import NavSection from "./Center"
import LogoSection from "./Left"
import MenuIconsSection from "./Right"

const HeaderContent = () => {
  const { toggleColorMode: toggleMode } = useColorMode()
  const text = useColorModeValue("dark", "light")
  const SwitchIcon = useColorModeValue(FaMoon, FaSun)
  return (
    <Flex w="100%" h="100%" px="6" align="center" justify="space-between">
      <LogoSection />
      <NavSection />
      <MenuIconsSection />

      <IconButton
        size="md"
        fontSize="lg"
        aria-label={`Switch to ${text} mode`}
        variant="ghost"
        color="current"
        ml="3"
        onClick={toggleMode}
        icon={<SwitchIcon />}
      />
    </Flex>
  )
}

const Header = () => {
  const bg = useColorModeValue("white", "#202020")
  return (
    <chakra.header
      pos="fixed"
      top="0"
      zIndex="1"
      bg={bg}
      left="0"
      right="0"
      boxShadow="base"
      width="full"
    >
      <chakra.div height="3.5rem" mx="auto" maxW="1200px">
        <HeaderContent />
      </chakra.div>
    </chakra.header>
  )
}

export default Header
