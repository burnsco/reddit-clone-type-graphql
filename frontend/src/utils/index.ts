export { default as formatCommentText } from "./formatCommentText"
export { default as customRender } from "./test-utils"
export { default as timeDifferenceForDate } from "./timeDifferenceForDate"
export { default as convertToErrorMap } from "./toErrorMap"
