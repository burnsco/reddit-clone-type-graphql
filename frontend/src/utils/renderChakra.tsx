import { ChakraProvider, CSSReset } from "@chakra-ui/core"
import theme from "@chakra-ui/theme"
import { render, RenderOptions } from "@testing-library/react"
import * as React from "react"

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <ChakraProvider theme={theme}>
    <CSSReset />
    {children}
  </ChakraProvider>
)

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options })

export { customRender as render }
