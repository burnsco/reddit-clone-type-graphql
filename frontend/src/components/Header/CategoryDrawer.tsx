import { useCreateSubredditMutation } from "@/generated/graphql"
import { useIsAuth } from "@/utils/useIsAuth"
import { gql } from "@apollo/client"
import {
  Alert,
  AlertIcon,
  Button,
  Drawer,
  DrawerBody,
  DrawerCloseButton,
  DrawerContent,
  DrawerFooter,
  DrawerHeader,
  DrawerOverlay,
  useDisclosure
} from "@chakra-ui/core"
import { Form, Formik } from "formik"
import { useRef } from "react"
import * as Yup from "yup"
import { ChakraField } from "../shared/ChakraField"

interface CreateSubredditProps {
  name: string
}

function CreateCategoryDrawer() {
  const { isOpen, onOpen, onClose } = useDisclosure()

  useIsAuth()
  const [
    submitSubreddit,
    { loading: mutationLoading, error: mutationError }
  ] = useCreateSubredditMutation()

  const btnRef = useRef<HTMLButtonElement | null>(null)

  if (mutationLoading) return null
  if (mutationError) {
    return (
      <Alert status="error">
        <AlertIcon />
        {mutationError.message}
      </Alert>
    )
  }

  const handleSubmit = (values: CreateSubredditProps) => {
    submitSubreddit({
      variables: {
        data: {
          name: values.name
        }
      },
      update(cache, { data }) {
        cache.modify({
          fields: {
            categories(existingCategories = []) {
              const newCategoryRef = cache.writeFragment({
                data: data?.createCategory.category,
                fragment: gql`
                  fragment NewCategory on Category {
                    id
                    name
                  }
                `
              })
              return [newCategoryRef, ...existingCategories]
            }
          }
        })
      }
    })
  }

  return (
    <>
      <Button ref={btnRef} size="md" colorScheme="blue" onClick={onOpen}>
        Subreddit
      </Button>
      <Drawer
        isOpen={isOpen}
        placement="top"
        onClose={onClose}
        finalFocusRef={btnRef}
      >
        <DrawerOverlay />
        <DrawerContent>
          <DrawerCloseButton />
          <DrawerHeader>Subreddit Name: </DrawerHeader>
          <Formik
            initialValues={{ name: "" }}
            validationSchema={Yup.object().shape({
              name: Yup.string().required("Required")
            })}
            onSubmit={(values, actions) => {
              setTimeout(() => {
                actions.setSubmitting(false)
                handleSubmit(values)
              }, 1000)
            }}
          >
            {({ isSubmitting, handleSubmit }) => (
              <Form onSubmit={handleSubmit}>
                <DrawerBody>
                  <ChakraField id="name" name="name" type="text" label="Name" />
                </DrawerBody>

                <DrawerFooter>
                  <Button variant="outline" mr={3} onClick={onClose}>
                    Cancel
                  </Button>
                  <Button type="submit" isLoading={isSubmitting} color="blue">
                    Submit
                  </Button>
                </DrawerFooter>
              </Form>
            )}
          </Formik>
        </DrawerContent>
      </Drawer>
    </>
  )
}

export default CreateCategoryDrawer
