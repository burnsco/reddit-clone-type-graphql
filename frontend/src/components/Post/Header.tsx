import { DeletePostDialog } from "@/components/shared/DeletePostDialog"
import { NextChakraLink } from "@/components/shared/NextChakraLink"
import { timeDifferenceForDate } from "@/utils/timeDifferenceForDate"
import {
  Box,
  Flex,
  HStack,
  IconButton,
  Spacer,
  useColorModeValue
} from "@chakra-ui/core"
import { useRouter } from "next/router"
import { FiEdit } from "react-icons/fi"
import PostCategory from "./Category"

const PostHeader: React.FC<{
  category?: string | null
  author?: string | null
  createdAt?: string | null
  postId?: string | null | undefined
}> = ({ category, author, createdAt, postId }) => {
  const fontColor = useColorModeValue("#1A1A1B", "gray.200")
  const router = useRouter()
  return (
    <HStack fontSize="sm" my={1} color={fontColor} w="full">
      <HStack>
        <PostCategory category={category} />
        <Box ml="2" textDecoration="none">
          Posted by{" "}
          <Box
            onClick={() => router.push(`/user/${author}`)}
            fontWeight="500"
            display="inline"
            color="gray.400"
            _hover={{
              textDecoration: "underline",
              cursor: "pointer"
            }}
          >
            {author}
          </Box>
          <Box display="inline" ml="2">
            {timeDifferenceForDate(Number(createdAt))}
          </Box>
        </Box>
      </HStack>
      <Spacer />
      {postId && (
        <Flex mr={1}>
          <NextChakraLink href="/post/edit/[id]" as={`/post/edit/${postId}}`}>
            <IconButton
              mr={2}
              size="xs"
              aria-label="Edit Post"
              icon={<FiEdit />}
            />
          </NextChakraLink>

          <DeletePostDialog postId={postId} category={category} />
        </Flex>
      )}
    </HStack>
  )
}

export default PostHeader
