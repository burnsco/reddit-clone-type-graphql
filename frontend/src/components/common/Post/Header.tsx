import { DeletePostDialog } from "@/components/common/DeletePostDialog"
import { NextChakraLink } from "@/components/common/index"
import { timeDifferenceForDate } from "@/utils/index"
import {
  Box,
  Flex,
  HStack,
  IconButton,
  Spacer,
  useColorModeValue
} from "@chakra-ui/core"
import { useRouter } from "next/router"
import { FiEdit } from "react-icons/fi"

const PostHeader: React.FC<{
  category?: string | null
  author?: string | null
  createdAt?: string | null
  postId?: string | null | undefined
}> = ({ category, author, createdAt, postId }) => {
  const fontColor = useColorModeValue("#1A1A1B", "gray.200")
  const router = useRouter()
  return (
    <HStack fontSize="sm" my={1} color={fontColor} w="full">
      <HStack>
        <Box
          fontWeight="600"
          color="orange.500"
          _hover={{
            textDecoration: "underline"
          }}
        >
          <Box
            _hover={{
              textDecoration: "underline",
              cursor: "pointer"
            }}
            onClick={() => router.push(`/r/${category}`)}
          >
            /r/{category}
          </Box>
        </Box>
        <Box ml="2" textDecoration="none">
          Posted by{" "}
          <Box
            onClick={() => router.push(`/user/${author}`)}
            fontWeight="500"
            display="inline"
            color="gray.400"
            _hover={{
              textDecoration: "underline",
              cursor: "pointer"
            }}
          >
            {author}
          </Box>
          <Box display="inline" ml="2">
            {timeDifferenceForDate(Number(createdAt))}
          </Box>
        </Box>
      </HStack>
      <Spacer />
      {postId && (
        <Flex mr={1}>
          <NextChakraLink href="/post/edit/[id]" as={`/post/edit/${postId}}`}>
            <IconButton
              mr={2}
              size="xs"
              aria-label="Edit Post"
              icon={<FiEdit />}
            />
          </NextChakraLink>

          <DeletePostDialog postId={postId} />
        </Flex>
      )}
    </HStack>
  )
}

export default PostHeader
