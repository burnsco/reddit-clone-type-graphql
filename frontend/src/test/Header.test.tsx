import { gql, InMemoryCache } from "@apollo/client"
import { MockedProvider } from "@apollo/client/testing"
import React from "react"
import Header from "../components/Header"
import { render } from "../utils/renderChakra"

const cache = new InMemoryCache()
cache.writeQuery({
  query: gql`
    query MeQuery {
      me {
        id
        username
      }
    }
  `,
  data: {
    me: {
      __typename: "User",
      id: "1",
      username: "Corey"
    }
  }
})

describe("Index", () => {
  it("Renders without Crashing", async () => {
    render(
      <MockedProvider cache={cache}>
        <Header />
      </MockedProvider>
    )
  })
})
