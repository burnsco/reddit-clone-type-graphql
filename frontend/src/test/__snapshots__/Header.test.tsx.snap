// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="css-11e9v45"
  >
    <nav
      class="css-iujbr9"
    >
      <button
        aria-controls="menu-list-2"
        aria-expanded="false"
        aria-haspopup="menu"
        class="chakra-button chakra-menu__menu-button css-18gjhjh"
        id="menu-button-2"
        type="button"
      >
        <span
          class="css-a9hmjk"
        >
          Register
        </span>
      </button>
      <button
        aria-controls="menu-list-2"
        aria-expanded="false"
        aria-haspopup="menu"
        class="chakra-button chakra-menu__menu-button css-1erotde"
        id="menu-button-2"
        type="button"
      >
        <span
          class="css-a9hmjk"
        >
          Login
        </span>
      </button>
      <button
        aria-label="Switch to light mode"
        class="chakra-button css-sx0t2q"
        type="button"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
          />
        </svg>
      </button>
    </nav>
  </header>
</DocumentFragment>
`;
